sql/src/main/java/io/crate/executor/transport/TransportShardUpdateAction.java

```diff
@@ -0,0 +1,335 @@
+        }
+
+        updateSourceByPaths(updatedSourceAsMap, pathsToUpdate);
+
+        final IndexRequest indexRequest = Requests.indexRequest(request.index()).type(request.type()).id(request.id()).routing(routing).parent(parent)
+                .source(updatedSourceAsMap, updateSourceContentType)
+                .version(getResult.getVersion());
+        indexRequest.operationThreaded(false);
+        return indexRequest;
+    }
+
+    /**
+     * Overwrite given values on the source. If the value is a map,
+     * it will not be merged but overwritten. The keys of the changes map representing a path of
+     * the source map tree.
+     * If the path doesn't exists, a new tree will be inserted.
+     *
+     * TODO: detect NOOP
+     */
+    private void updateSourceByPaths(Map<String, Object> source, Map<String, Object> changes) {
+        for (Map.Entry<String, Object> changesEntry : changes.entrySet()) {
+            if (changesEntry.getKey().contains(".")) {
+                // sub-path detected, dive recursive to the wanted tree element
+                List<String> path = Splitter.on(".").splitToList(changesEntry.getKey());
+                String currentKey = path.get(0);
+                if (!source.containsKey(currentKey)) {
+                    // insert parent tree element
+                    source.put(currentKey, new HashMap<String, Object>());
+                }
+                Map<String, Object> subChanges = new HashMap<>();
+                subChanges.put(Joiner.on(".").join(path.subList(1, path.size())),
+                        changesEntry.getValue());
+                updateSourceByPaths((Map<String, Object>) source.get(currentKey), subChanges);
+            } else {
+                // overwrite or insert the field
+                source.put(changesEntry.getKey(), changesEntry.getValue());
+            }
+        }
+    }
+
+
+    static FieldExtractor<GetResult> buildExtractor(final String field) {
+        return new FieldExtractor<GetResult>() {
+            @Override
+            public Object extract(GetResult getResult) {
+                assert getResult.sourceAsMap() != null;
+                return XContentMapValues.extractValue(field, getResult.sourceAsMap());
+            }
+        };
+    }
+
+    static class Context {
+        private final List<String> fields;
+        private final Functions functions;
+        private String[] fieldsArray;
+
+        Context(Functions functions, int size) {
+            this.functions = functions;
+            fields = new ArrayList<>(size);
+        }
+
+        public void addField(String fieldName) {
+            fields.add(fieldName);
+        }
+
+        public String[] fields() {
+            if (fieldsArray == null) {
+                fieldsArray = fields.toArray(new String[fields.size()]);
+            }
+            return fieldsArray;
+        }
+    }
+
+    static class Visitor extends SymbolVisitor<Context, FieldExtractor<GetResult>> {
```

Comment:
try to build a shared  `FieldExtractorBuilder<T>`  and use it here and in ESGetTask